File: app/src/main/java/com/egecius/gtdx/db/DbImpl.java
package com.egecius.gtdx.db;

import com.egecius.gtdx.datatypes.TodoTask;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.Map;
import java.util.UUID;

import rx.Observable;
import rx.Subscriber;

/**
 * Database implemented by Firebase
 */
public final class DbImpl implements Db {

	private static final String TASKS = "tasks";
	/** last id used for a task */
	private static final String LAST_ID = "last_id";

	private final DatabaseReference refRoot;
	private final DatabaseReference refTasks;

	public DbImpl(final FirebaseDatabase firebaseDatabase) {
		refRoot = firebaseDatabase.getReference();
		refTasks = refRoot.child(TASKS);
	}

	@Override
	public void addTask(final TodoTask task) {
		refTasks.child(createId()).setValue(task);
	}

	private String createId() {
		return UUID.randomUUID().toString();
	}

	@Override
	public Observable<Map<String, Map>> getAllTasks() {
		return Observable.create(new Observable.OnSubscribe<Map<String, Map>>() {
			@Override
			public void call(final Subscriber<? super Map<String, Map>> subscriber) {
				refTasks.addValueEventListener(new ValueEventListener() {
					@Override
					public void onDataChange(final DataSnapshot dataSnapshot) {
						//noinspection unchecked
						Map<String, Map> map = (Map<String, Map>) dataSnapshot.getValue();
						subscriber.onNext(map);
					}

					@Override
					public void onCancelled(final DatabaseError databaseError) {
						subscriber.onError(new DbExcpetion(databaseError));
					}
				});
			}
		});
	}
}

File: app/src/main/kotlin/com/egecius/gtdx/db/DbImpl.kt
package com.egecius.gtdx.db

import com.egecius.gtdx.datatypes.TodoTask
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import java.util.UUID

import rx.Observable
import rx.Subscriber

/**
 * Database implemented by Firebase
 */
class DbImpl(firebaseDatabase: FirebaseDatabase) : Db {

    private val refRoot: DatabaseReference = firebaseDatabase.reference
    private val refTasks: DatabaseReference

    init {
        refTasks = refRoot.child(TASKS)
    }

    override fun addTask(taskTitle: String) {
        val task = TodoTask(createId(), taskTitle, System.currentTimeMillis())
        refTasks.child(task.id).setValue(task)
    }

    override fun getContexts(): Observable<List<String>> {
        throw UnsupportedOperationException("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    private fun createId(): String {
        return UUID.randomUUID().toString()
    }

    override val getAllTasks: Observable<Map<String, Map<*, *>>>
        get() = Observable.create { subscriber ->
            refTasks.addValueEventListener(object : ValueEventListener {
                override fun onDataChange(dataSnapshot: DataSnapshot) {
                    val map = dataSnapshot.value as Map<String, Map<*, *>>
                    subscriber.onNext(map)
                }

                override fun onCancelled(databaseError: DatabaseError) {
                    subscriber.onError(DbException(databaseError))
                }
            })
        }

    override fun removeTask(id: String) {
        refTasks.child(id).removeValue()
    }

    companion object {
        private val TASKS = "tasks"
    }
}
